API Pull - Add an API to return Google WPCOM App auth URL

ianlin opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Part of #2146. 

This PR adds an API which returns an auth URL for granting access to Google WPCOM App.

The format of the auth URL:

```
https://public-api.wordpress.com/oauth2/authorize?
client_id=CLIENT_ID
&redirect_uri=PARTNER_REDIRECT_URL
&response_type=code
&blog=BLOD_ID
&scope=wc-partner-access
&state=BASE64_ENCODED_STRING
```

The content of the `state` is an URL query string where the value of its parameter `redirect_url` is being URL encoded. E.g.
`nonce=nonce-123&redirect_url=https%3A%2F%2Fmerchant-site.example.com%2Fwp-admin%2Fadmin.php%3Fpage%3Dwc-admin%26path%3D%2Fgoogle%2Fsetup-mc`

and the decoded version is:

`nonce=nonce-123&redirect_url=https://merchant-site.example.com/wp-admin/admin.php?page=wc-admin&path=/google/setup-mc`

where the `redirect_url` will be used to return back to the merchant's site by the partner.

Project: pcTzPl-1SM-p2
WPCOM App OAuth: https://developer.wordpress.com/docs/oauth2/

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Request to `GET /wc/gla/rest-api/authorize`.
2. Check the returned auth URL has the same format from the description above.
3. Open this URL in the browser, you should see the permission granting page.
4. Make sure you can actually grant the permission and it will be redirected to woo.com.
5. After redirected to woo.com, copy the value of the `state` URL query parameter, which is a base64 encoded string.
6. Go to https://www.base64decode.org/ and decode base64 string.
7. The decoded value should look like
    ```
    nonce=nonce-123&redirect_url=https%3A%2F%2Fmerchant-site.example.com%2Fwp-admin%2Fadmin.php%3Fpage%3Dwc-admin%26path%3D%2Fgoogle%2Fsetup-mc
    ```

### Additional details:

Note that this PR hardcoded three things in the code at the moment as we will be calling a new Google API to get those data when the API is ready:

**Client ID**

WPCOM app Client ID, currently I'm using the WPCOM app I created, it should be Google's client ID in the future.

**Redirect URI**

An URL that Google will set in their WPCOM app settings. I'm currently using https://woo.com set in my WPCOM app.

**nonce**

A validation nonce that would return by the new Google API as well.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Add - API to return Google WPCOM auth URL

ianlin committed:
Add an API to get Google WPCOM auth URL

ianlin committed:
Add unit tests for RestAPI/AuthController

ianlin commented on src/API/Site/Controllers/RestAPI/AuthController.php:
```diff
@@ -0,0 +1,181 @@
+	 *
+	 * @var string[]
+	 */
+	private const NEXT_PATH_MAPPING = [
+		'setup-mc'  => '/google/setup-mc',
+		'reconnect' => '/google/settings',
+	];
```
Adding this because in GL&A plugin, both onboarding flow and the settings page will call this API so we need to know where we are redirecting the merchant back.

mikkamp commented on src/API/Site/Controllers/RestAPI/AuthController.php:
```diff
@@ -0,0 +1,181 @@
+
+			try {
+				$auth_url = 'https://public-api.wordpress.com/oauth2/authorize';
+
+				// TODO: Call an API by Google with merchant information and get the below data.
+				// We'd probably need use WCS to communicate with the new API.
+				$client_id    = '91299';
+				$redirect_uri = 'https://woo.com';
+				$nonce        = 'nonce-123';
+
+				$response_type = 'code';
+				$blog_id       = Jetpack_Options::get_option( 'id' );
+				$scope         = 'wc-partner-access';
+
+				$next                  = $request->get_param( 'next_page_name' );
+				$path                  = self::NEXT_PATH_MAPPING[ $next ];
+				$merchant_redirect_url = urlencode_deep( admin_url( "admin.php?page=wc-admin&path={$path}" ) );
+
+				$state = base64_encode(
+					build_query(
+						[
+							'nonce'        => $nonce,
+							'redirect_url' => $merchant_redirect_url,
+						]
+					)
+				);
+
+				$auth_url = esc_url_raw(
+					add_query_arg(
+						[
+							'blog'          => $blog_id,
+							'client_id'     => $client_id,
+							'redirect_uri'  => $redirect_uri,
+							'response_type' => $response_type,
+							'scope'         => $scope,
+							'state'         => $state,
+						],
+						$auth_url
+					)
+				);
```
Not sure if this PR is fully ready for a review yet, as a lot of it seems preliminary. However it looks like this controller is getting a lot of responsibility to build an auth URL. Can we move that to a service class which can just have the function of building an auth URL (at least for now, until we need more there). I think it will make it easier for unit testing and allow setting contants within the class without them interfering with the controller.

The path mapping could still remain in the controller if we want to match the Google connect controller. But the controller should not care much else about how that URL is built. Especially if we have it send an external request to fetch some of the URL parts.

I'm not sure what folder services like this should end up. But maybe since it's related to the WordPress API. We could consider adding it in `src/API/WP`. The `NotificationService` seems to have been added to `src/Google` but maybe we'd want to consider having them grouped together, since they both do similar things using the blog_id and handling the URL to `public-api.wordpress.com`.

ianlin replied:
Thanks for the suggestion @mikkamp.

> Not sure if this PR is fully ready for a review yet, as a lot of it seems preliminary

Since this PR is going to be merged into a feature branch rather than `develop` so I was thinking using the hardcoded data first for testing and revisit it when we get the information about the new Google API.

> Can we move that to a service class which can just have the function of building an auth URL

Completely agreed, why didn't I do it... 😓 Working on it.

> We could consider adding it in src/API/WP. The NotificationService seems to have been added to src/Google but maybe we'd want to consider having them grouped together, since they both do similar things using the blog_id and handling the URL to public-api.wordpress.com.

Great idea, thanks.

ianlin committed:
Merge branch 'feature/google-api-project' into add/get-google-wpcom-app-authorize-api

ianlin committed:
Extract responsibility from AuthController to a new OAuthService

ianlin committed:
Update unit test for RestAPI/AuthController

ianlin committed:
Add tests for WP/OAuthService

ianlin committed:
Move NotificationsService to the folder API/WP

ianlin committed:
Fix PHPCS

ianlin commented:
Created an `OAuthService` and added the corresponding tests in d4c0099, b3ff1e6, and ff1a516 regarding the comment https://github.com/woocommerce/google-listings-and-ads/pull/2291#discussion_r1512788358.

Moved `NotificationsService` under the folder `API/WP` in 7950c95. cc @puntope since you're the author of `NotificationsService`.

mikkamp commented on src/API/WP/OAuthService.php:
```diff
@@ -0,0 +1,96 @@
+
+		$merchant_redirect_url = urlencode_deep( admin_url( "admin.php?page=wc-admin&path={$path}" ) );
+
+		$state = base64_encode(
```
Just a quick comment, we should be making sure this is URL safe, maybe some helper functions like:
https://base64.guru/developers/php/examples/base64url
https://www.php.net/manual/en/function.base64-encode.php#103849

ianlin committed:
Add helper functions base64url_encode/base64url_decode

ianlin committed:
Remove functions return union types

As PHP 7.4 does not support it and we still support 7.4

ianlin committed:
Merge branch 'feature/google-api-project' into add/get-google-wpcom-app-authorize-api

jorgemd24 commented on src/API/Site/Controllers/RestAPI/AuthController.php:
```diff
@@ -0,0 +1,132 @@
+			]
+		);
+	}
+
```
Shouldn't we link the schema with the endpoint? `'schema' => $this->get_api_response_schema_callback()`

ianlin replied:
Updated in 94886fc.

jorgemd24 commented on src/API/WP/OAuthService.php:
```diff
@@ -0,0 +1,99 @@
+
+namespace Automattic\WooCommerce\GoogleListingsAndAds\API\WP;
+
+use Automattic\Jetpack\Connection\Client;
```
Unused dependency.

ianlin replied:
Updated in efb5e5f.

jorgemd24 commented on src/API/WP/OAuthService.php:
```diff
@@ -0,0 +1,99 @@
+	 * where its URL decoded version is:
+	 * nonce=nonce-123&redirect_url=https://merchant-site.example.com/wp-admin/admin.php?page=wc-admin&path=/google/setup-mc
+	 *
+	 * @param string $path An URL parameter for the path within GL&A page, which will be added in the merchant redirect URL.
```
I think we need to update the docs to tell that the state will be base64 url encoded.

ianlin replied:
Updated in 70442c0.

jorgemd24 commented on src/API/WP/OAuthService.php:
```diff
@@ -0,0 +1,99 @@
+	 *
+	 * @return array An associative array contains required information from Google.
+	 */
+	protected function get_data_from_google(): array {
```
We could explain the response properties, for example for the nonce we could say that it is fetched from Google and it will be used to verify the call.

ianlin replied:
Updated in 9c70657.

jorgemd24 commented on src/API/WP/OAuthService.php:
```diff
@@ -0,0 +1,99 @@
+	protected function get_data_from_google(): array {
+		return [
+			'client_id'    => '91299',
+			'redirect_uri' => 'https://woo.com',
```
Also worth noting, is that if the redirect_url will be fetched from the Google side, they need to be sure that it matches with the WPCOM Client settings. The redirect URLs only work if they are added in WPCOM.

ianlin replied:
Yeah, I will make sure Google is aware of this.

jorgemd24 commented on src/HelperTraits/Utilities.php:
```diff
@@ -78,4 +78,46 @@ protected function gla_setup_for( $seconds ): bool {
+		$b64 = base64_encode( $data );
+
+		// Make sure you get a valid result, otherwise, return FALSE, as the base64_encode() function do
+		if ( $b64 === false ) {
```
We say that `base64_encode` can return a boolean `false` but reading the doc it seems that always will be a string: https://www.php.net/manual/en/function.base64-encode.php

ianlin replied:
You're right, it's `base64_decode` function would return false, not `base64_encode`.

Updated in 614df20.

jorgemd24 commented on tests/Unit/API/Site/Controllers/RestAPI/AuthControllerTest.php:
```diff
@@ -0,0 +1,79 @@
+			->willReturn( $expected_auth_url );
+
+		$response = $this->do_request( self::ROUTE_AUTHORIZE, 'GET' );
+		$data     = $response->get_data();
```
Unused variable.

ianlin replied:
Updated in 2b6f251.

jorgemd24 commented on src/API/WP/OAuthService.php:
```diff
@@ -0,0 +1,99 @@
+
+		$merchant_redirect_url = urlencode_deep( admin_url( "admin.php?page=wc-admin&path={$path}" ) );
+
+		$state = $this->base64url_encode(
```
What happens if `base64url_encode` returns `false`?

ianlin replied:
This is not the case because `base64url_encode` won't return `false` anymore. Updated in 614df20.

jorgemd24 reviewed (commented):
Hi @ianlin! Thanks for working on this! It is working well and I am getting the right URL. I just left some comments about the code and some questions.

ianlin committed:
Link response schema to /rest-api/authorize endpoint

ianlin committed:
Remove unused dependency in WP/OAuthService

ianlin committed:
Update the PHPDoc to tell the state will be url safe base64 encoded

ianlin committed:
Update PHPDoc to explain the response data from Google

ianlin committed:
Do not return false because base64_encode always returns string

ianlin committed:
Remove unused variable

ianlin commented:
Thanks @jorgemd24 for the review and nice suggestions. It's ready for the second round.

jorgemd24 commented on src/HelperTraits/Utilities.php:
```diff
@@ -78,4 +78,41 @@ protected function gla_setup_for( $seconds ): bool {
+	 * @return string
+	 */
+	protected function base64url_encode( $data ): string {
+		// First of all you should encode $data to Base64 string
```
I think this comment could be removed, since the `base64_encode` function is self-explanatory.

jorgemd24 commented on src/API/Site/Controllers/RestAPI/AuthController.php:
```diff
@@ -0,0 +1,133 @@
+		return function ( Request $request ) {
+			try {
+				$next     = $request->get_param( 'next_page_name' );
+				$path     = self::NEXT_PATH_MAPPING[ $next ];
```
WDYT if we move `self::NEXT_PATH_MAPPING` to the OAuthService? This way, the controller only needs to pass the parameter, and the OAuthService handles the logic to determine the correct path.

ianlin replied:
I put `self::NEXT_PATH_MAPPING` in the controller to match other controllers like [API/Site/Controllers/Google/AccountController](https://github.com/woocommerce/google-listings-and-ads/blob/292cf662bfc6bf976ed926785bdd86cbca175631/src/API/Site/Controllers/Google/AccountController.php#L104) and [API/Site/Controllers/Jetpack/AccountController](https://github.com/woocommerce/google-listings-and-ads/blob/292cf662bfc6bf976ed926785bdd86cbca175631/src/API/Site/Controllers/Jetpack/AccountController.php#L126). I think it's fine for controllers doing such thing like request transformation. If we would move it to the service layer then it means the service layer would need to understand about the request of controller. Which to me is a little bit not that "separation of concerns".

jorgemd24 commented on src/API/Site/Controllers/RestAPI/AuthController.php:
```diff
@@ -0,0 +1,133 @@
+	 */
+	private const NEXT_PATH_MAPPING = [
+		'setup-mc'  => '/google/setup-mc',
+		'reconnect' => '/google/settings',
```
What do you think if we change `reconnect` with `settings`? Mainly because the query param is `next_page_name` so I think it makes sense to call it 'settings'.

ianlin replied:
Yeah, I think changing to `settings` sounds more straightforward to me. I updated in b41bad3, however in Google/Jetpack's AccountController they still use reconnect as well as some places in the frontend, I will create a follow up issue so we could do it in another PR.

jorgemd24 reviewed (approved):
Thanks @ianlin for the adjustments. I left some comments about the code and a suggestion of moving `self::NEXT_PATH_MAPPING to the OAuthService` that I would like to hear your opinion! 

I am approving the PR as it works as expected 👍

ianlin committed:
Remove an unnecessary comment since the code is self-explanatory

ianlin committed:
Rename reconnect to settings for page mapping
